lib/cpu_config.rb

```diff
@@ -0,0 +1,161 @@
+require_relative 'benchmark_runner'
+
+# Manages CPU frequency and turbo boost configuration for benchmark consistency
+class CPUConfig
+  class << self
+    # Configure CPU for benchmarking: disable frequency scaling and verify settings
+    def configure_for_benchmarking(turbo:)
+      build.configure_for_benchmarking(turbo: turbo)
+    end
+
+    def build
+      if File.exist?(IntelCPUConfig::PSTATE_DIR)
+        IntelCPUConfig.new
+      elsif File.exist?(AMDCPUConfig::BOOST_PATH)
+        AMDCPUConfig.new
+      else
+        NullCPUConfig.new
+      end
+    end
+  end
+
+  def configure_for_benchmarking(turbo:)
+    disable_frequency_scaling(turbo: turbo)
+    check_pstate(turbo: turbo)
+  end
+
+  private
+
+  def disable_frequency_scaling(turbo:)
+    disable_turbo_boost unless turbo || turbo_disabled?
+    maximize_frequency unless frequency_maximized?
+  end
+
+  def turbo_disabled?
+    # Override in subclasses
+    false
+  end
+
+  def frequency_maximized?
+    # Override in subclasses
+    false
+  end
+
+  def disable_turbo_boost
+    # Override in subclasses
+  end
+
+  def maximize_frequency
+    # Override in subclasses
+  end
+
+  def check_pstate(turbo:)
+    # Override in subclasses
+  end
+end
+
+# Intel CPU configuration
+class IntelCPUConfig < CPUConfig
+  PSTATE_DIR = '/sys/devices/system/cpu/intel_pstate'
+  NO_TURBO_PATH = "#{PSTATE_DIR}/no_turbo"
+  MIN_PERF_PCT_PATH = "#{PSTATE_DIR}/min_perf_pct"
+  TURBO_DISABLED_VALUE = '1'
+  FREQUENCY_MAXIMIZED_VALUE = '100'
+
+  private
+
+  def disable_turbo_boost
+    # sudo requires the flag '-S' in order to take input from stdin
+    BenchmarkRunner.check_call("sudo -S sh -c 'echo #{TURBO_DISABLED_VALUE} > #{NO_TURBO_PATH}'")
+    at_exit { BenchmarkRunner.check_call("sudo -S sh -c 'echo 0 > #{NO_TURBO_PATH}'", quiet: true) }
+  end
+
+  def maximize_frequency
+    # Disabling Turbo Boost reduces the CPU frequency, so this should be run after that.
```

Comment:
I feel like this comment should actually go in `disable_frequency_scaling` since that's where the order is important.